devices.go

```diff
@@ -172,10 +183,48 @@ func (d *deviceCache) SetDeviceByIdOpState(id string, os models.OperatingState)
 
 // Update updates the device in the cache and ensures that the
 // copy in Core Metadata is also updated.
-func (d *deviceCache) Update(id string) error {
+func (d *deviceCache) Update(dev *models.Device) error {
+	addressable, err := updateAddressable(&dev.Addressable)
+	if err != nil {
+		return err
+	}
+	dev.Addressable = *addressable
+
+	device, err := svc.dc.DeviceForName(dev.Name)
+	if err != nil {
+		return err
+	}
+
+	dev.Id = device.Id
+	err = svc.dc.Update(*dev)
+	if err != nil {
+		return err
+	}
+
+	d.devices[dev.Name] = dev
 	return nil
 }
 
+// updateAddressable updates the addressable.
+// Because addressable client doesn't have update function. If addressable not found then add it.
+func updateAddressable(addr *models.Addressable) (*models.Addressable, error) {
+	addressable, err := svc.ac.AddressableForName(addr.Name)
+
+	if err != nil && addr.Name != "" && strings.Contains(err.Error(), "Item not found") {
+		id, err := svc.ac.Add(addr)
+		if err != nil {
+			return addr, err
+		}
+		addr.Id = bson.ObjectIdHex(id)
+		return addr, nil
+
+	} else if err != nil {
+		return addr, err
+	}
+
+	return &addressable, nil
```

Comment:
if the Addressable exists, it still needs to return the new addressable, not the old one